core/src/main/java/org/infinispan/interceptors/distribution/BaseDistributionInterceptor.java

```diff
@@ -493,32 +601,76 @@ private Object handleLocalOnlyReadManyCommand(InvocationContext ctx, VisitableCo
-                                                             InvocationContext ctx, C command,
-                                                             MergingCompletableFuture<Object> allFuture,
-                                                             int destinationIndex,
-                                                             ReadManyCommandHelper<C> helper) {
-      ReplicableCommand remoteCommand = helper.copyForRemote(command, keys, ctx);
-      rpcManager.invokeRemotelyAsync(Collections.singleton(owner), remoteCommand, defaultSyncOptions)
-         .whenComplete((responseMap, throwable) -> {
-            if (throwable != null) {
-               allFuture.completeExceptionally(throwable);
-            }
-            Response response = getSingleSuccessfulResponseOrFail(responseMap, allFuture);
-            if (response == null) return;
-            try {
-               Object responseValue = ((SuccessfulResponse) response).getResponseValue();
-               Object[] values = unwrapFunctionalManyResultOnOrigin(ctx, keys, responseValue);
-               if (values != null) {
-                  System.arraycopy(values, 0, allFuture.results, destinationIndex, values.length);
-                  allFuture.countDown();
-               } else {
-                  allFuture.completeExceptionally(new IllegalStateException("Unexpected response value " + responseValue));
+   private class ReadManyHandler<C extends TopologyAffectedCommand> implements BiConsumer<Map<Address, Response>, Throwable> {
+      private final MergingCompletableFuture<Object> allFuture;
+      private final InvocationContext ctx;
+      private final C command;
+      private final List<Object> keys;
+      private final int destinationIndex;
+      private final Collection<Address> contactedNodes;
+      private final ReadManyCommandHelper<C> helper;
+
+      private ReadManyHandler(MergingCompletableFuture<Object> allFuture, InvocationContext ctx, C command, List<Object> keys, Collection<Address> contactedNodes, int destinationIndex, ReadManyCommandHelper<C> helper) {
+         this.allFuture = allFuture;
+         this.ctx = ctx;
+         this.command = command;
+         this.keys = keys;
+         this.destinationIndex = destinationIndex;
+         this.contactedNodes = contactedNodes;
+         this.helper = helper;
+      }
+
+      @Override
+      public void accept(Map<Address, Response> responseMap, Throwable throwable) {
+         if (throwable != null) {
+            allFuture.completeExceptionally(throwable);
+            return;
+         }
+         Response response = getSingleReadResponseOrFail(responseMap, allFuture);
+         if (response == null) {
+            if (!allFuture.isDone()) {
+               try {
+                  handleMissingResponse();
+               } catch (Throwable t) {
+                  allFuture.completeExceptionally(t);
```

Comment:
Right, the `handleMissingResponse()` method is different... still, IntelliJ is telling me the code is duplicated :)

It would be nice to have a single place where that comment is required, but I wouldn't mind if you just copy it here.